src/main/java/com/aws/greengrass/detector/uploader/ConnectivityUpdater.java

```diff
@@ -76,7 +77,7 @@ synchronized void uploadAddresses(List<String> ips, Config config) {
                 logger.atInfo().kv("IPs", ips).kv("defaultPort", defaultPort).log("Uploaded IP addresses");
             }
         } catch (GreengrassV2DataException | SdkClientException e) {
-            Throwable cause = e.getCause();
+            Throwable cause = Utils.getUltimateCause(e);
```

Comment:
Two questions:
1. Has `getUltimateCause` been around for a while? Want to make sure we're not breaking compatibility with older versions
2. Is `SdkClientException` appropriate or should we be catching the more general `SdkException`?

Reply:
1. It was there in the first Nucleus

Reply:
2. `ggDataClient.updateConnectivityInfo()` throws `SdkClientException`. Though not necessary, its worth catching more general `SdkException` as we would be covering both client and server side exception scenarios:
- when service could not be contacted for a response,
- when client is unable to parse the response from service
- when client's request was correctly transmitted to the service, but for some reason, the service was not able to process it.

Let me update the PR.

Reply:
`SdkException` catch covers `GreengrassV2DataException`

Reply:
Thanks. This is much better. One more follow up, in the spirit of the original PR. We should be able to also specifically catch a 403 and log about the iot policy in that case. Else, "we don't know, here's the stack trace"

Reply:
That makes sense. Updated